File: ts-sdk/src/auth/device_flow.ts
// OAuth 2.0 Device Authorization Grant (RFC 8628) client — the browser
// `tn auth login` flow. Idiomatic with `az login` / `gh auth login`:
//
//   1. POST /api/v1/device/code  → { device_code, user_code, verification_uri,
//      verification_uri_complete, interval, expires_in }. The request is SIGNED
//      by the device key so the vault enrolls a DID the caller provably owns.
//   2. The CLI opens `verification_uri_complete` (best-effort) AND always prints
//      the short `user_code` + `verification_uri` so a non-opening browser is a
//      non-event — sign in on any device.
//   3. POST /api/v1/device/token  (poll with `device_code`) → RFC error bodies
//      `authorization_pending` / `slow_down` until approved, then 200
//      `{ account_id, did }`. The device key stays the principal — no token to
//      store; enrollment is the result.
//
// The wire shapes here are the contract of record; the vault implements the
// same field + error names.

import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { Buffer } from "node:buffer";

import type { DeviceKey } from "../core/signing.js";

/** Successful `/device/code` response (RFC 8628 §3.2 fields, snake_case wire). */
export interface DeviceCodeResponse {
  /** Opaque secret the CLI polls the token endpoint with. */
  deviceCode: string;
  /** Short, human-typeable code (e.g. `WDJB-MJHT`) the user enters in the browser. */
  userCode: string;
  /** The stable URL to visit (e.g. `https://vault…/device`). */
  verificationUri: string;
  /** `verification_uri` with `user_code` pre-filled — for the auto-open path. */
  verificationUriComplete: string;
  /** Minimum seconds between token polls. */
  interval: number;
  /** Seconds until the device_code expires. */
  expiresIn: number;
}

/** Device-flow failure. `code` carries the RFC 8628 error slug when the vault
 *  returned one (`expired_token`, `access_denied`, …). */
export class DeviceFlowError extends Error {
  readonly code: string | null;
  constructor(message: string, code: string | null = null) {
    super(message);
    this.name = "DeviceFlowError";
    this.code = code;
  }
}

const DEFAULT_HEADERS: Record<string, string> = { "Content-Type": "application/json" };

/** Standard (not url-safe) base64 — what the Python vault's `b64decode` expects. */
function toStandardBase64(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64");
}

/** The message the device key signs to prove DID ownership at `/device/code`:
 *  SHA-256 of `tn:device-code:<did>`. */
function deviceCodeMessage(did: string): Uint8Array {
  return new Uint8Array(createHash("sha256").update(`tn:device-code:${did}`, "utf8").digest());
}

/**
 * Start a device-authorization request. Signs `tn:device-code:<did>` with the
 * device key so the vault binds the pending login to a DID the caller owns.
 */
export async function requestDeviceCode(
  vaultBase: string,
  deviceKey: DeviceKey,
  opts: { fetchImpl?: typeof fetch; label?: string } = {},
): Promise<DeviceCodeResponse> {
  const fetchImpl = opts.fetchImpl ?? globalThis.fetch.bind(globalThis);
  const base = vaultBase.replace(/\/+$/, "");
  const did = deviceKey.did;
  const signatureB64 = toStandardBase64(deviceKey.sign(deviceCodeMessage(did)));

  const body: Record<string, unknown> = { did, signature_b64: signatureB64 };
  if (opts.label) body["label"] = opts.label;

  const resp = await fetchImpl(`${base}/api/v1/device/code`, {
    method: "POST",
    headers: DEFAULT_HEADERS,
    body: JSON.stringify(body),
  });
  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new DeviceFlowError(
      `device/code returned ${resp.status}: ${text.slice(0, 256)}`,
    );
  }
  const j = (await resp.json()) as Record<string, unknown>;
  const deviceCode = j["device_code"];
  const userCode = j["user_code"];
  const verificationUri = j["verification_uri"];
  if (typeof deviceCode !== "string" || typeof userCode !== "string" || typeof verificationUri !== "string") {
    throw new DeviceFlowError(`device/code response missing fields: ${JSON.stringify(j)}`);
  }
  return {
    deviceCode,
    userCode,
    verificationUri,
    verificationUriComplete:
      typeof j["verification_uri_complete"] === "string"
        ? (j["verification_uri_complete"] as string)
        : `${verificationUri}?code=${encodeURIComponent(userCode)}`,
    interval: typeof j["interval"] === "number" ? j["interval"] : 5,
    expiresIn: typeof j["expires_in"] === "number" ? j["expires_in"] : 900,
  };
}

/** Result of a completed device-flow login. */
export interface DeviceTokenResult {
  accountId: string;
  did: string;
  /** Base64 key-id for a pre-sealed AWK pickup, or null if the vault didn't
   *  include one (older vault or no AWK enrolled yet). */
  awkPickupKeyId: string | null;
}

/**
 * Poll `/device/token` until the user approves in the browser, the code
 * expires, or approval is denied. Honors RFC 8628 `authorization_pending`
 * (keep waiting) and `slow_down` (add 5s to the interval). `sleep` is injectable
 * for tests; a monotonic clock isn't needed — the loop bounds itself on
 * `expiresIn` poll-budget so a stubbed sleep can't spin forever.
 */
export async function pollDeviceToken(
  vaultBase: string,
  dc: DeviceCodeResponse,
  opts: { fetchImpl?: typeof fetch; sleep?: (ms: number) => Promise<void> } = {},
): Promise<DeviceTokenResult> {
  const fetchImpl = opts.fetchImpl ?? globalThis.fetch.bind(globalThis);
  const sleep = opts.sleep ?? ((ms: number) => new Promise<void>((r) => setTimeout(r, ms)));
  const base = vaultBase.replace(/\/+$/, "");
  let interval = dc.interval;
  let elapsed = 0;

  for (;;) {
    if (elapsed >= dc.expiresIn) {
      throw new DeviceFlowError("device code expired before approval", "expired_token");
    }
    await sleep(interval * 1000);
    elapsed += interval;

    const resp = await fetchImpl(`${base}/api/v1/device/token`, {
      method: "POST",
      headers: DEFAULT_HEADERS,
      body: JSON.stringify({ device_code: dc.deviceCode }),
    });

    if (resp.status === 200) {
      const j = (await resp.json()) as Record<string, unknown>;
      const accountId = j["account_id"];
      const did = j["did"];
      if (typeof accountId !== "string" || !accountId) {
        throw new DeviceFlowError(`device/token ok but missing account_id: ${JSON.stringify(j)}`);
      }
      return {
        accountId,
        did: typeof did === "string" ? did : "",
        awkPickupKeyId: typeof j["awk_pickup_key_id"] === "string" ? (j["awk_pickup_key_id"] as string) : null,
      };
    }

    // RFC 8628 §3.5: pending/slow_down are normal; everything else is terminal.
    let errCode = "";
    try {
      errCode = String(((await resp.json()) as Record<string, unknown>)["error"] ?? "");
    } catch {
      // non-JSON error body — treat as terminal below
    }
    if (errCode === "authorization_pending") continue;
    if (errCode === "slow_down") {
      interval += 5;
      continue;
    }
    if (errCode === "expired_token") {
      throw new DeviceFlowError("device code expired before approval", "expired_token");
    }
    if (errCode === "access_denied") {
      throw new DeviceFlowError("sign-in was denied in the browser", "access_denied");
    }
    throw new DeviceFlowError(
      `device/token returned ${resp.status}${errCode ? ` (${errCode})` : ""}`,
      errCode || null,
    );
  }
}

/**
 * Best-effort browser open (the auto path). Returns true if a launcher was
 * spawned, false if none is available — the caller ALWAYS prints the code + URL
 * regardless, so a false here is a non-event. Never throws.
 */
export function openBrowser(url: string): boolean {
  // Headless / no display → don't even try; the printed URL is the path.
  if (process.env["TN_NO_BROWSER"] === "1") return false;
  const plat = process.platform;
  let cmd: string;
  let args: string[];
  if (plat === "win32") {
    // `start` is a cmd builtin; the empty "" is the window-title arg.
    cmd = "cmd";
    args = ["/c", "start", "", url];
  } else if (plat === "darwin") {
    cmd = "open";
    args = [url];
  } else {
    cmd = "xdg-open";
    args = [url];
  }
  try {
    const child = spawn(cmd, args, { stdio: "ignore", detached: true });
    child.on("error", () => {
      /* launcher missing — best-effort, the printed URL covers it */
    });
    child.unref();
    return true;
  } catch {
    return false;
  }
}

File: python/tn/device_flow.py
"""OAuth 2.0 Device Authorization Grant (RFC 8628) client — the browser
``tn auth login`` flow. Cross-impl parity with ``ts-sdk/src/auth/device_flow.ts``
(identical wire field + error names).

Idiomatic with ``az login`` / ``gh auth login``: open the browser AND print a
short, human-typeable code as the fallback. The device key stays the principal —
on success the vault returns ``{account_id, did}``; there is no token to store.

Contract:

    POST /api/v1/device/code  {did, signature_b64}
        signature over SHA-256("tn:device-code:<did>") — proves DID ownership.
        -> {device_code, user_code, verification_uri,
            verification_uri_complete, interval, expires_in}
    POST /api/v1/device/token {device_code}   (poll)
        -> 400 {error: authorization_pending | slow_down | expired_token
                | access_denied}  until approved, then 200 {account_id, did}
"""

from __future__ import annotations

import base64
import hashlib
import os
import time
import webbrowser
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

import httpx

_USER_AGENT = "tn-proto-device-flow"
_DEFAULT_TIMEOUT = 30.0


class DeviceFlowError(Exception):
    """Device-flow failure. ``code`` carries the RFC 8628 error slug when the
    vault returned one (``expired_token`` / ``access_denied`` / ...)."""

    def __init__(self, message: str, code: str | None = None) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class DeviceCode:
    device_code: str
    user_code: str
    verification_uri: str
    verification_uri_complete: str
    interval: int
    expires_in: int


def _device_code_message(did: str) -> bytes:
    return hashlib.sha256(f"tn:device-code:{did}".encode()).digest()


def request_device_code(
    vault_base: str,
    sk: Any,
    did: str,
    *,
    label: str | None = None,
    client: httpx.Client | None = None,
) -> DeviceCode:
    """Start a device-authorization request. ``sk`` is an Ed25519 private key
    (``cryptography`` ``Ed25519PrivateKey``); the request is signed so the vault
    enrolls a DID the caller provably owns."""
    base = vault_base.rstrip("/")
    sig_b64 = base64.b64encode(sk.sign(_device_code_message(did))).decode("ascii")
    body: dict[str, Any] = {"did": did, "signature_b64": sig_b64}
    if label:
        body["label"] = label

    owns = client is None
    c = client or httpx.Client(timeout=_DEFAULT_TIMEOUT, headers={"User-Agent": _USER_AGENT})
    try:
        resp = c.post(f"{base}/api/v1/device/code", json=body)
    finally:
        if owns:
            c.close()
    if resp.status_code != 200:
        raise DeviceFlowError(f"device/code returned {resp.status_code}: {resp.text[:256]}")
    j = resp.json()
    user_code = j["user_code"]
    verification_uri = j["verification_uri"]
    return DeviceCode(
        device_code=j["device_code"],
        user_code=user_code,
        verification_uri=verification_uri,
        verification_uri_complete=j.get("verification_uri_complete")
        or f"{verification_uri}?code={user_code}",
        interval=int(j.get("interval", 5)),
        expires_in=int(j.get("expires_in", 900)),
    )


def poll_device_token(
    vault_base: str,
    dc: DeviceCode,
    *,
    client: httpx.Client | None = None,
    sleep: Callable[[float], None] = time.sleep,
) -> dict[str, str]:
    """Poll ``/device/token`` until the user approves, the code expires, or it's
    denied. Honors RFC 8628 ``authorization_pending`` (wait) / ``slow_down``
    (add 5s). Returns ``{"account_id", "did"}``. ``sleep`` is injectable for
    tests; the loop bounds itself on the poll budget so a stubbed sleep can't
    spin forever."""
    base = vault_base.rstrip("/")
    interval = dc.interval
    elapsed = 0

    owns = client is None
    c = client or httpx.Client(timeout=_DEFAULT_TIMEOUT, headers={"User-Agent": _USER_AGENT})
    try:
        while True:
            if elapsed >= dc.expires_in:
                raise DeviceFlowError("device code expired before approval", "expired_token")
            sleep(interval)
            elapsed += interval

            resp = c.post(f"{base}/api/v1/device/token", json={"device_code": dc.device_code})
            if resp.status_code == 200:
                j = resp.json()
                account_id = j.get("account_id")
                if not account_id:
                    raise DeviceFlowError(f"device/token ok but missing account_id: {j!r}")
                return {"account_id": account_id, "did": j.get("did", ""),
                        "awk_pickup_key_id": j.get("awk_pickup_key_id")}

            err = ""
            try:
                err = str(resp.json().get("error", ""))
            except Exception:  # noqa: BLE001 — non-JSON body → terminal below
                pass
            if err == "authorization_pending":
                continue
            if err == "slow_down":
                interval += 5
                continue
            if err == "expired_token":
                raise DeviceFlowError("device code expired before approval", "expired_token")
            if err == "access_denied":
                raise DeviceFlowError("sign-in was denied in the browser", "access_denied")
            raise DeviceFlowError(
                f"device/token returned {resp.status_code}" + (f" ({err})" if err else ""),
                err or None,
            )
    finally:
        if owns:
            c.close()


def open_browser(url: str) -> bool:
    """Best-effort browser open (the auto path). Returns True if a launcher was
    invoked. The caller ALWAYS prints the code + URL regardless, so False is a
    non-event. ``TN_NO_BROWSER=1`` skips. Never raises."""
    if os.environ.get("TN_NO_BROWSER") == "1":
        return False
    try:
        return webbrowser.open(url)
    except Exception:  # noqa: BLE001 — best-effort; the printed URL covers it
        return False
